### ISSUE TITLE
Add Windows Certificate Store Option for Local Machine (instead of Current User)

### ISSUE THREAD

**Issue by dharders** [electron-userland/electron-builder#1717]
* **Version**:  19.6.1

* **Target**: Windows NSIS (win 10 x64)

I use the certificate store on Windows when code-signing, so I don't have to specify CSC_LINK/PASSWORD etc as environment variables and can just have my build config as 
```javascript
{
"win": {
      "target": [
        "nsis"
      ],
      "certificateSubjectName": "My Company"
    }
}
```

This has worked fine in Win7, however recently I purchased a new Win10 PC and imported my certificate then ran a build and received the error `SignTool Error: No certificates were found that met all the given criteria.`. The cert was imported correctly and had the correct user permissions .

After a bit of debugging I found that on Windows 10, if you add your code signing certificate (.pfx in my case) using the "Manage Computer Certificates" in Control Panel->Administrative Tools, the certificate lands in the LOCAL MACHINE store by default and not the CURRENT USER store.

electron builder, uses `signtool.exe` without the `/sm` flag, so by default it looks in the CURRENT USER store. 

```
C:\Users\User\AppData\Local\electron-builder\cache\winCodeSign\winCodeSign-1.7.0\windows-10\x64\signtool.exe sign /t "http://timestamp.verisign.com/scripts/timstamp.dll" /n "My Company" /du "https://www.mywebsite.com.au" "C:\source\app\package_desktop\win-unpacked\App.exe"
SignTool Error: No certificates were found that met all the given criteria.
```

Can I suggest adding a `{ win:  { useLocalMachineStore: true } }` option or similar, so new win 10 users don't get as confused as I did, or maybe update the docs to emphasize that the CURRENT USER store is used instead (or console warning) ?

P.S. fix is: add the following to [windowsCodeSign.ts#L137](https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/src/windowsCodeSign.ts#L137)

```javascript
if (isWin && options.useLocalMachineStore) {
    args.push("/sm");
}
```

or is there already a command line arg that we can pass directly to `signtool` params ?

**Comment by dharders** [comment]
I'm thinking of submitting a PR for this feature @develar .

However, since `signtool.exe` has many options, should we just create a single master option of, say, `win.customCodeSignArgs` to cover any future user requests also ?

If `win.customCodeSignArgs` is defined, the user takes full responsibility as this string would be passed verbatim to `signtool.exe`  thus it would  **replace** the existing args (except the last file input path arg), i.e. user must also specify timestamp etc. in the `win.customCodeSignArgs` string option.

Another option is to have it just **add** to the existing args, but this is a bit more complex as it involves string parsing and arg replacement, which could be error prone.

I think the first option is best, as it is the most flexible and if you were using this option then you most likely already know what to specify when code signing on Windows.

If neither options are satisfactory, I could submit a PR of the original `win.useLocalMachineStore` boolean option still ?

Keen to hear your thoughts @develar before I continue...

**Comment by develar** [comment]
Thanks for investigation. Well — no option should be added at all. electron-builder should try to find cert in the current user first, and, if not found, in the local store. Using "retry with flag" or some smart solution (if there is a way to check where cert located).


Bloody windows — it is MS fault, not outs or user's. We need hide this from user.

**Comment by dharders** [comment]
Yes that would be a better solution. Less config is always a win!

I guess you could just splice the `/sm` flag into the args array on a retry (if failed).

I might leave this one to you to implement as my retry logic probably won't be as robust as your solution. Thanks for all that you do in this repo @develar !!!

**Comment by develar** [comment]
Moved to backlog to keep issue list clear.

**Comment by develar** [comment]
19.33.0

**Comment by dharders** [comment]
Great work as always! Thanks a million ;)

**Comment by dharders** [comment]
Just noticed a regression on v19.33.0 @develar 

I've been using Win10 store, with standard cert (i.e. **Non EV**) with following options:
```javascript
{
"win": {
      "target": [
        "nsis"
      ],
      "certificateSubjectName": "My Company"
    }
}
```
and assuming I had my cert in the correct store, elec-builder would find it.

However, your new feature should **auto find** the cert, which is great! However, it couldn't find my (previously) working cert. A bit of debugging found that [windowsCodeSign.ts#L99](https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/src/windowsCodeSign.ts#L99) is comparing ISSUER Name to Subject name, i.e.

```typescript
if (!certInfo.IssuerName.Name.includes(certificateSubjectName))
```
where I think it should be
```typescript
if (!certInfo.Subject.includes(certificateSubjectName))
```

Logging the certInfo object shows that IssuerName only contains `IssuerName:
   { Name: 'CN=COMODO RSA Code Signing CA, O=COMODO CA Limited, L=Salford, S=Greater Manchester, C=GB' ... `. This may be due to it being a standard cert and not EV (I don't have an EV to test).

So in summary, `v19.33.0` I get codesign error. With my above suggested change it works.

Also note, if I don't specify `certificateSubjectName` and use `certificateSha1` instead it works with `v19.33.0`.
